File: app/src/main/java/com/anglersparadise/ui/lake/LakeViewModel.java
// app/src/main/java/com/anglersparadise/ui/lake/LakeViewModel.java
package com.anglersparadise.ui.lake;

import android.os.Handler;
import android.os.Looper;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.anglersparadise.data.FishRepository;
import com.anglersparadise.domain.model.Fish;
import com.anglersparadise.domain.model.SpeciesCatalog;
import com.anglersparadise.util.SingleLiveEvent;
import java.util.Random;

public class LakeViewModel extends ViewModel {

    private final MutableLiveData<LakeState> _state = new MutableLiveData<>(LakeState.IDLE);
    public LiveData<LakeState> state = _state;

    private final SingleLiveEvent<String> _toast = new SingleLiveEvent<>();
    public LiveData<String> toast = _toast;

    private final Handler handler = new Handler(Looper.getMainLooper());
    private Runnable waitTask = null;
    private final Random rng = new Random();

    public void cast() {
        if (_state.getValue() != LakeState.IDLE) return;
        _state.setValue(LakeState.WAITING);

        cancelWait();
        long delayMs = 1500 + rng.nextInt(2501); // 1500..4000
        waitTask = () -> {
            if (_state.getValue() == LakeState.WAITING) {
                _state.setValue(LakeState.HOOKED);
                _toast.setValue("Fish on!  Reel now!");
            }
        };
        handler.postDelayed(waitTask, delayMs);
    }

    public void reel() {
        LakeState s = _state.getValue();
        if (s == LakeState.HOOKED) {
            _state.setValue(LakeState.CAUGHT);
            _toast.setValue("You caught a fish.");
        } else if (s == LakeState.WAITING) {
            _state.setValue(LakeState.ESCAPED);
            _toast.setValue("Too early—fish got away.");
            handler.postDelayed(this::resetToIdle, 1200);
        }
    }

    public void confirmCatchToTank() {
        if (_state.getValue() == LakeState.CAUGHT) {
            String species = SpeciesCatalog.randomName();
            int size = SpeciesCatalog.randomSize();
            FishRepository.addToTank(FishRepository.newFish(species, size));
            _toast.setValue("Added a " + species + " (size " + size + ") to tank.");
            resetToIdle();
        }
    }

    public void letGo() {
        if (_state.getValue() == LakeState.CAUGHT) resetToIdle();
    }

    private void resetToIdle() {
        cancelWait();
        _state.setValue(LakeState.IDLE);
    }

    private void cancelWait() {
        if (waitTask != null) {
            handler.removeCallbacks(waitTask);
            waitTask = null;
        }
    }

    @Override protected void onCleared() {
        super.onCleared();
        cancelWait();
    }
}

File: app/src/main/java/com/anglersparadise/ui/lake/LakeViewModel.kt
// app/src/main/java/com/anglersparadise/ui/lake/LakeViewModel.kt

package com.anglersparadise.ui.lake

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.random.Random

enum class LakeState { IDLE, WAITING, HOOKED, CAUGHT, ESCAPED }

class LakeViewModel : ViewModel() {

    private val _state = MutableStateFlow(LakeState.IDLE)
    val state = _state.asStateFlow()

    private val _toast = MutableSharedFlow<String>(extraBufferCapacity = 4)
    val toast = _toast.asSharedFlow()

    private var waitJob: Job? = null

    fun cast() {
        if (_state.value != LakeState.IDLE) return
        _state.value = LakeState.WAITING

        waitJob?.cancel()
        waitJob = viewModelScope.launch {
            val delayMs = Random.nextLong(1500L, 4000L)
            delay(delayMs)
            if (_state.value == LakeState.WAITING) {
                _state.value = LakeState.HOOKED
                _toast.tryEmit("Fish on!  Reel now!")
            }
        }
    }

    fun reel() {
        when (_state.value) {
            LakeState.HOOKED -> {
                _state.value = LakeState.CAUGHT
                _toast.tryEmit("You caught a fish.")
            }
            LakeState.WAITING -> {
                _state.value = LakeState.ESCAPED
                _toast.tryEmit("Too early—fish got away.")
                viewModelScope.launch {
                    delay(1200)
                    resetToIdle()
                }
            }
            else -> Unit
        }
    }

    fun confirmCatchToTank() {
        if (_state.value == LakeState.CAUGHT) {
            val species = com.anglersparadise.domain.model.SpeciesCatalog.randomName()
            val size = com.anglersparadise.domain.model.SpeciesCatalog.randomSize()
            com.anglersparadise.data.FishRepository.addToTank(
                com.anglersparadise.data.FishRepository.newFish(species, size)
            )
            _toast.tryEmit("Added a $species (size $size) to tank.")
            resetToIdle()
        }
    }

    fun letGo() {
        if (_state.value == LakeState.CAUGHT) {
            resetToIdle()
        }
    }

    private fun resetToIdle() {
        waitJob?.cancel()
        _state.value = LakeState.IDLE
    }
}
